### PR TITLE
Update predictor to properly present archive status.

### PR DESCRIPTION
**Pull request by anoto-moniz**
In the near future, we'll be adding support for predictor versions. There will be a separation between a predictor version and the predictor root, which can be archived independently. However, the predictor root will be invisible to the user. Since this logic already exists in the backend, that means the value of "is_archived" may no longer match the expected value.

Until we expose that separation and make this explicit, we'll overwrite the archive status to present the root's status, which matches what the user previously expected. In addition, we point users at the new methods, which act on the predictor root.

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Maintenance (non-breaking change to assist developers)

### Adherence to team decisions
- [X] I have added tests for 100% coverage
- [X] I have written Numpy-style docstrings for every method and class.
- [ ] I have communicated the downstream consequences of the PR to others.
- [X] I have bumped the version in __version\__.py

### CODE REVIEW

File: src/citrine/__version__.py [hunk 1]
```diff
@@ -1 +1 @@
-__version__ = '1.38.0'
+__version__ = '1.38.1'
```
Comment (sfriedowitz) [inline src/citrine/__version__.py:1]: Think this should be a minor b/c adding new `archive_root` and `restore_root` methods.

File: src/citrine/resources/predictor.py [hunk 2]
```diff
@@ -96,38 +128,69 @@ def update(self, predictor: Predictor) -> Predictor:
...
+        return any(uid == archived_pred.uid for archived_pred in self.list_archived())
+
+    @deprecated(deprecated_in="1.38.1", removed_in="2.0.0",
+                details="archive() is deprecated in favor of archive_root().")
+    def archive(self, uid: Union[UUID, str]) -> Predictor:
+        """[DEPRECATED] Archive a root predictor."""
+        self.archive_root(uid)
+        return self.get(uid)
+
+    @deprecated(deprecated_in="1.38.1", removed_in="2.0.0",
```
Comment (sfriedowitz) [inline src/citrine/resources/predictor.py:169]: Update deprecated calls if changing version to a minor.

**Review by sfriedowitz — COMMENTED** [review]
looks good, just a version comment.

File: src/citrine/__version__.py (hunk 1 above)
Comment (anoto-moniz) [reply src/citrine/__version__.py:1]: That's a fair point.

File: src/citrine/resources/predictor.py [hunk 3]
```diff
@@ -52,20 +54,50 @@ def _predictors_path(self, subpath: str, uid: Union[UUID, str] = None):
         path_template = self._path_template + (f"/{uid}" if uid else "") + f"/{subpath}"
         return format_escaped_url(path_template, project_id=self.project_id)
 
+    # Until we expose predictor versions in the SDK, and thus expose the
+    # difference between archving a predictor root and version, we should
+    # report the root's archive status.
```
Comment (kroenlein) [inline src/citrine/resources/predictor.py:59]: Should this be labeled with a `TODO`?

File: src/citrine/resources/predictor.py [hunk 4]
```diff
@@ -52,20 +54,50 @@ def _predictors_path(self, subpath: str, uid: Union[UUID, str] = None):
...
+    def _inject_archive_info(self, predictor, is_archived=None):
```
Comment (kroenlein) [inline src/citrine/resources/predictor.py:60]:
```suggestion
    def _inject_archive_info(self, predictor: Predictor, is_archived: bool = None) -> Predictor:
```
It's internal, but typehints are still convenient

File: src/citrine/resources/predictor.py [hunk 5]
```diff
@@ -52,20 +54,50 @@ def _predictors_path(self, subpath: str, uid: Union[UUID, str] = None):
...
+        if archived:
+            predictor.archived_by = predictor.updated_by
+            predictor.archive_time = predictor.update_time
+        else:
+            predictor.archived_by = None
+            predictor.archive_time = None
+        return predictor
+
+    # Same as above
+    def _build_collection_elements(self, collection, is_archived=None):
```
Comment (kroenlein) [inline src/citrine/resources/predictor.py:72]:
```suggestion
    def _build_collection_elements(self, collection: PredictorCollection, is_archived: bool = None) -> Predictor:
```
and so on...

File: src/citrine/resources/predictor.py [hunk 6]
```diff
@@ -96,38 +128,69 @@ def update(self, predictor: Predictor) -> Predictor:
...
+        """Archive a root predictor.
 
         uid: Union[UUID, str]
             Unique identifier of the predictor to archive.
 
         """
         path = self._predictors_path("archive", uid)
-        entity = self.session.put_resource(path, {}, version=self._api_version)
-        return self.build(entity)
+        self.session.put_resource(path, {}, version=self._api_version)
```
Comment (kroenlein) [inline src/citrine/resources/predictor.py:142]: I don't necessarily agree with always returning an object, but you are changing the pattern here.

File: src/citrine/resources/predictor.py [hunk 7]
```diff
@@ -96,38 +128,69 @@ def update(self, predictor: Predictor) -> Predictor:
...
+        """Restore an archived root predictor.
 
         uid: Union[UUID, str]
             Unique identifier of the predictor to restore.
 
         """
         path = self._predictors_path("restore", uid)
-        entity = self.session.put_resource(path, {}, version=self._api_version)
-        return self.build(entity)
+        self.session.put_resource(path, {}, version=self._api_version)
```
Comment (kroenlein) [inline src/citrine/resources/predictor.py:152]: Same as above

File: tests/resources/test_predictor.py [hunk 8]
```diff
@@ -50,48 +54,58 @@ def test_delete():
 
 
 def test_archive_and_restore(valid_label_fractions_predictor_data):
-    session = mock.Mock()
+    session = FakeSession()
```
Comment (kroenlein) [inline tests/resources/test_predictor.py:57]: Thank you for continuing this migration

File: tests/resources/test_predictor.py [hunk 9]
```diff
@@ -50,48 +54,58 @@ def test_delete():
...
     pc = PredictorCollection(uuid.uuid4(), session)
     entity = deepcopy(valid_label_fractions_predictor_data)
     entity["metadata"]["archived"] = entity["metadata"]["created"]
-    session.put_resource.return_value = entity
+    session.set_responses(None, entity, paging_response(entity))
 
-    archived_predictor = pc.archive(uuid.uuid4())
+    with pytest.deprecated_call():
+        archived_predictor = pc.archive(uuid.uuid4())
```
Comment (kroenlein) [inline tests/resources/test_predictor.py:64]: It seems like you are missing the chance to also test the new method directly.  I understand that the test cascades, but the person following after you may not grasp that.

**Review by kroenlein — COMMENTED** [review]
Generally good.  Some minor questions and tweaks.

File: src/citrine/resources/predictor.py (hunk 3 above)
Comment (anoto-moniz) [reply src/citrine/resources/predictor.py:59]: Sure.

File: src/citrine/resources/predictor.py (hunk 4 above)
Comment (anoto-moniz) [reply src/citrine/resources/predictor.py:60]: Fair enough.

File: tests/resources/test_predictor.py (hunk 9 above)
Comment (anoto-moniz) [reply tests/resources/test_predictor.py:64]: Good point. I'll break them out.

File: src/citrine/resources/predictor.py (hunk 6 above)
Comment (anoto-moniz) [reply src/citrine/resources/predictor.py:142]: That's part of why it's a new method.

**Review by kroenlein — APPROVED** [review]
LGTM

**Pull request merged**